packages/xo-web/src/common/modal.js

```diff
@@ -259,57 +260,108 @@ export const confirm = ({ body, icon = 'alarm', title, strongConfirm }) =>
 
 // -----------------------------------------------------------------------------
 
-const preventDefault = event => event.preventDefault()
-
-class FormModal extends BaseComponent {
-  state = {
-    value: this.props.defaultValue,
-  }
-
-  get value () {
-    return this.state.value
-  }
-
-  render () {
-    const { body, formId } = this.props
-    return (
-      <form id={formId} onSubmit={preventDefault}>
-        {cloneElement(body, {
-          value: this.state.value,
-          onChange: this.linkState('value'),
-        })}
-      </form>
-    )
-  }
-}
-
-export const form = ({ body, defaultValue, icon, title, size }) => {
-  const formId = generateRandomId()
-  const buttons = [
-    {
-      btnStyle: 'primary',
-      label: _('formOk'),
-      form: formId,
-    },
-  ]
-  return new Promise((resolve, reject) => {
-    modal(
-      <GenericModal
-        buttons={buttons}
-        icon={icon}
-        reject={reject}
-        resolve={resolve}
-        title={title}
-      >
-        <FormModal body={body} defaultValue={defaultValue} formId={formId} />
-      </GenericModal>,
+let formModalState
+export const form = props =>
+  new Promise((resolve, reject) => {
+    formModalState.props = {
       reject,
-      {
-        bsSize: size,
-      }
-    )
+      resolve,
+      showModal: true,
+      ...props,
+    }
+    disableShortcuts()
   })
-}
+
+const getInitialState = () => ({
+  id: generateRandomId(),
+  localValue: undefined,
+  props: undefined,
+})
+export const FormModal = [
+  provideState({
+    initialState: getInitialState,
+    effects: {
+      initialize () {
+        if (formModalState !== undefined) {
+          throw new Error('FormModal is a singleton!')
+        }
+        formModalState = this.state
+      },
+      finalize: () => {
+        formModalState = undefined
+      },
+      close: () => () => {
+        enableShortcuts()
+        return getInitialState()
+      },
+      cancel ({ close }) {
+        this.state.props.reject()
+        close()
+      },
+      submit ({ close }) {
+        const {
+          props: { resolve },
+          value,
+        } = this.state
+        resolve(value)
+        close()
+      },
+      onChange: (_, localValue) => () => ({
+        localValue,
+      }),
+    },
+    computed: {
+      value: ({ props = {}, localValue = props.defaultValue }) => localValue,
+    },
+  }),
+  injectState,
+  ({ state: { props = {}, ...state }, effects }) => (
```

Comment:
It's a bad idea to clone the `state` object, it forces all computed properties to be evaluated and mark all values in the state as dependencies (which triggers a re-render if changed).